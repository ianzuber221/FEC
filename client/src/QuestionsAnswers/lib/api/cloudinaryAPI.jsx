import axios from 'axios';

export default function getImageUrl(file) {
  const postImageConfig = {
    method: 'POST',
    url: 'https://api.cloudinary.com/v1_1/dl9zxpaoq/upload',
    data: {
      file,
      upload_preset: 'upload_preset_atelier',
    },
  };

  return axios(postImageConfig)
    .then((result) => result.data.url)
    .catch((err) => {
      console.log('error uploading images', err);
    });
}
