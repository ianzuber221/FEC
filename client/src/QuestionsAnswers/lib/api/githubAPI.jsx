import axios from 'axios';

export function getProductInfo(productId) {
  const requestConfig = {
    method: 'GET',
    url: `${process.env.API_URL}/products/${productId}`,
    headers: {
      Authorization: process.env.AUTH_KEY,
    },
  };

  return axios(requestConfig)
    .catch((err) => {
      console.log('failed fetching product info.', err);
    });
}

export function getQuestions(productId, page = 1, count = 200) {
  const requestConfig = {
    method: 'GET',
    url: `${process.env.API_URL}/qa/questions`,
    params: {
      product_id: productId,
      count,
      page,
    },
    headers: {
      Authorization: process.env.AUTH_KEY,
    },
  };
  return axios(requestConfig)
    .catch((err) => {
      console.log('failed fetching all questions from API.', err);
    });
}

export function postQuestion(productId, { body, name, email }) {
  const requestConfig = {
    method: 'POST',
    url: `${process.env.API_URL}/qa/questions`,
    data: {
      body,
      name,
      email,
      product_id: Number(productId),
    },
    headers: {
      Authorization: process.env.AUTH_TOKEN,
    },
  };

  return axios(requestConfig)
    .catch((err) => {
      console.log('error posting question', err);
    });
}

export function getAnswers(questionId, page = 1, count = 200) {
  const requestConfig = {
    method: 'GET',
    url: `${process.env.API_URL}/qa/questions/${questionId}/answers`,
    params: {
      page,
      count,
    },
    headers: {
      Authorization: process.env.AUTH_KEY,
    },
  };
  return axios(requestConfig)
    .catch((err) => {
      console.log('failed to get answers', err);
    });
}

export function postAnswer(questionId, { body, name, email, photos }) {
  const requestConfig = {
    method: 'POST',
    url: `${process.env.API_URL}/qa/questions/${questionId}/answers`,
    data: {
      body,
      name,
      email,
      photos,
    },
    headers: {
      Authorization: process.env.AUTH_KEY,
    },
  };
  return axios(requestConfig)
    .catch((err) => {
      console.log('failed posting answer.', err);
    });
}

export function updateHelpful(type, id) {
  let path;
  if (type === 'question') path = 'qa/questions';
  if (type === 'answer') path = 'qa/answers';
  if (type === 'review') path = 'review';

  const requestConfig = {
    method: 'PUT',
    url: `${process.env.API_URL}/${path}/${id}/helpful`,
    headers: {
      Authorization: process.env.AUTH_TOKEN,
    },
  };

  return axios(requestConfig)
    .catch((err) => {
      console.log(`Failed PUT request for marking ${type} of id ${id} helpful.`, err);
    });
}

export function updateReport(type, id) {
  let path;
  if (type === 'question') path = 'qa/questions';
  if (type === 'answer') path = 'qa/answers';

  console.log(path);
  console.log(type);

  const requestConfig = {
    method: 'PUT',
    url: `${process.env.API_URL}/${path}/${id}/report`,
    headers: {
      Authorization: process.env.AUTH_TOKEN,
    },
  };

  return axios(requestConfig)
    .catch((err) => {
      console.log('failed reporting answer', err);
    });
}