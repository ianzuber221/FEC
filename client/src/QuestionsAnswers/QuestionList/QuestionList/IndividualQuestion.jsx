import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import _ from 'underscore';
import { getAnswers, postAnswer } from '../../lib/api/githubAPI';
import getImageUrl from '../../lib/api/cloudinaryAPI';
import Helpful from '../../components/Helpful';
import Report from '../../components/Report';
import AddAnswerForm from '../AddAnswer/AddAnswerForm';
import AnswerList from '../AnswerList/AnswerList';

function IndividualQuestion({ productName, question, renderQuestions, keyword }) {
  const [showModal, setShowModal] = useState(false);
  const [answerList, setAnswerList] = useState(sortAnswers(question.answers));
  const [page, setPage] = useState(1);
  const [count, setCount] = useState(100);
  const [helpfulCount, setHelpfulCount] = useState(question.question_helpfulness);

  const renderAnswers = () => {
    getAnswers(question.question_id, page, count)
      .then((result) => {
        if (result.data.results.length === 0) {
          return;
        }
        setAnswerList(sortAnswers(result.data.results));
      });
  };

  const addAnswer = (formValues) => {
    if (formValues.photos) {
      const photoUrlPromise = formValues.photos.map((photo) => getImageUrl(photo.file))
      Promise.all(photoUrlPromise)
        .then((result) => {
          formValues.photos = result;
          postAnswer(question.question_id, formValues)
            .then(() => {
              renderAnswers();
            });
        });
    } else {
      postAnswer(question.question_id, formValues)
        .then(() => {
          renderAnswers();
        });
    }
  };

  useEffect(() => {
    if (answerList.length >= count) {
      setCount(count * 2);
    }
  }, [answerList]);

  return (
    <DivContainer show={question.question_body.match(new RegExp(keyword, 'i'))}>
      <DivQuestion>
        <QContainer>
          <Title>Q:</Title>
          <BoldBody>{question.question_body}</BoldBody>
        </QContainer>
        <OptionsDiv>
          <Helpful
            id={question.question_id}
            type="question"
            currentCount={helpfulCount}
            setHelpfulCount={setHelpfulCount}
            renderComponent={renderQuestions}
            tabIndex="0"
          />
          <PaddedSpan>{' | '}</PaddedSpan>
          <Report
            id={question.question_id}
            type="question"
          />
          <PaddedSpan>{' | '}</PaddedSpan>
          <ClickableU onClick={() => setShowModal(true)} onKeyDown={() => setShowModal(true)} role="button" tabIndex="-1">
            Add Answer
          </ClickableU>
        </OptionsDiv>
      </DivQuestion>
      <AnswerList
        answerList={answerList}
        renderAnswers={renderAnswers}
        Title={Title}
      />
      <AddAnswerForm
        question={question}
        submitHandler={addAnswer}
        productName={productName}
        show={showModal}
        setShowModal={setShowModal}
      />
    </DivContainer>
  );
}

export default IndividualQuestion;

const sortAnswers = (answers) => (
  _(answers).chain()
    .sortBy('helpfulness')
    .sortBy((answer) => answer.answerer_name.toLowerCase() === 'seller')
    .reverse()
    .value()
);

const DivContainer = styled.div`
  display: ${props => props.show ? 'block' : 'none'}
`;

const DivQuestion = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 15px 0;
  padding-right: 10px;
`;

const Title = styled.span`
  width: 30px;
  font-weight: bold;
`;

const QContainer = styled.div`
  display: flex;
  width: 60%;
`;

const BoldBody = styled.b`
  width: 100%;
`;

const OptionsDiv = styled.div`
  font-size: 11px;
  font-weight: 300;
  color: #77787a;
  display: flex;
  max-width: 40%;
  align-items: center;
`;

const PaddedSpan = styled.div`
  padding: 0 10px;
  font-size: 15px;
`;

const ClickableU = styled.u`
  &:hover {
    cursor: pointer;
    opacity: 60%;
  }
`;