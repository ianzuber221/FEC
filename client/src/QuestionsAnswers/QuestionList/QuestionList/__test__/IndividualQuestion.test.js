/**
 * @jest-environment jsdom
 */
/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { jest } from '@jest/globals';
import IndividualQuestion from '../IndividualQuestion';
import '@testing-library/jest-dom/extend-expect';
import * as apis from '../../../lib/api/githubAPI';

require('dotenv').config();

const renderMock = jest.fn();

it('should match snapshot', () => {
  const component = renderer.create(
    <IndividualQuestion
      productName="sample"
      question={{
        question_id: 123,
        question_helpfulness: 3,
        question_body: 'new answer',
      }}
      renderQuestions={renderMock}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});

it('should show modal upon clicking "add answer"', async () => {
  const { getByText, getByTestId } = render(
    <IndividualQuestion
      productName="sample"
      question={{
        question_id: 123,
        question_helpfulness: 3,
        question_body: 'new answer',
      }}
      renderQuestions={renderMock}
    />
  );

  const addAnswer = getByText('Add Answer');
  await userEvent.click(addAnswer);

  const modal = getByTestId('add-answer-modal');
  expect(modal).toBeInTheDocument();

});

it('should render after form is submitted', async () => {
  render(
    <IndividualQuestion
      productName="sample"
      question={{
        question_id: 543290,
        question_helpfulness: 3,
        question_body: 'new question',
        answers: [],
      }}
      renderQuestions={renderMock}
    />
  );

  const answer = screen.queryByText('Hi this is from the Seller');
  expect(answer).toBeNull();

  const addAnswer = screen.getByText('Add Answer');
  await userEvent.click(addAnswer);

  const modal = screen.getByTestId('add-answer-modal');
  expect(modal).toBeInTheDocument();

  const AnswerInput = screen.getByTestId('answer-input');
  const usernameInput = screen.getByTestId('username-input');
  const emailInput = screen.getByTestId('email-input');
  await userEvent.type(AnswerInput, 'new answer');
  await userEvent.type(usernameInput, 'new username');
  await userEvent.type(emailInput, 'email@gmail.com');

  const submitButton = screen.getByTestId('form-button-test');
  await userEvent.click(submitButton);

  const modal2 = screen.queryByTestId('add-answer-modal');
  expect(modal2).toBeNull();

  const answerText = await screen.findByText('Hi this is from the Seller');
  expect(answerText).toBeInTheDocument();

});
