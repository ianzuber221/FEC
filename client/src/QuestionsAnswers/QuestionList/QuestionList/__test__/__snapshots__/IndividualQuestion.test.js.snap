// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="individual-question"
>
  <div
    className="sc-llJcti fiutnT"
  >
    <div
      className="sc-gicCDI fdbmIx"
    >
      <span
        className="sc-iIPllB fXNfNz"
      >
        Q:
      </span>
      <b
        className="sc-ezWOiH esfWkg"
      >
        new answer
      </b>
    </div>
    <div
      className="css-1iivxpu-MuiStack-root"
    >
      <div
        className="sc-dkzDqf hoZXih helpful"
      >
        <span
          className="sc-bczRLJ grWmXc"
        >
          Helpful?
        </span>
        <u
          className="sc-gsnTZi hBfuwO"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="-1"
        >
          Yes
        </u>
        <span>
          (3)
        </span>
      </div>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
      />
      <u
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex="-1"
      >
        Add Answer
      </u>
    </div>
  </div>
  <div
    className="sc-bBrHrO ibWLUA"
  >
    <span
      className="sc-iIPllB fXNfNz"
    />
    <div>
      No answers.
    </div>
  </div>
</div>
`;
